services/discourse_translator/amazon.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module DiscourseTranslator
+  class Amazon < Base
+    require 'aws-sdk-translate'
+
+    MAXLENGTH = 5000
+
+    SUPPORTED_LANG = {
+      af: 'af',
+      sq: 'sq',
+      am: 'am',
+      ar: 'ar',
+      hy: 'hy',
+      az: 'az',
+      bn: 'bn',
+      bs: 'bs',
+      bg: 'bg',
+      ca: 'ca',
+      zh: 'zh',
+      zh_TW: 'zh-TW',
+      hr: 'hr',
+      cs: 'cs',
+      da: 'da',
+      fa_AF: 'fa-AF',
+      nl: 'nl',
+      en: 'en',
+      et: 'et',
+      fa: 'fa',
+      tl: 'tl',
+      fi: 'fi',
+      fr: 'fr',
+      fr_CA: 'fr-CA',
+      ka: 'ka',
+      de: 'de',
+      el: 'el',
+      gu: 'gu',
+      ht: 'ht',
+      ha: 'ha',
+      he: 'he',
+      hi: 'hi',
+      hu: 'hu',
+      is: 'is',
+      id: 'id',
+      ga: 'ga',
+      it: 'it',
+      ja: 'ja',
+      kn: 'kn',
+      kk: 'kk',
+      ko: 'ko',
+      lv: 'lv',
+      lt: 'lt',
+      mk: 'mk',
+      ms: 'ms',
+      ml: 'ml',
+      mt: 'mt',
+      mr: 'mr',
+      mn: 'mn',
+      no: 'no',
+      ps: 'ps',
+      pl: 'pl',
+      pt: 'pt',
+      pt_PT: 'pt-PT',
+      pa: 'pa',
+      ro: 'ro',
+      ru: 'ru',
+      sr: 'sr',
+      si: 'si',
+      sk: 'sk',
+      sl: 'sl',
+      so: 'so',
+      es: 'es',
+      es_MX: 'es-MX',
+      sw: 'sw',
+      sv: 'sv',
+      ta: 'ta',
+      te: 'te',
+      th: 'th',
+      tr: 'tr',
+      uk: 'uk',
+      ur: 'ur',
+      uz: 'uz',
+      vi: 'vi',
+      cy: 'cy'
+    }
+
+    def self.access_token_key
+      "aws-translator"
+    end
+
+    def self.detect(post)
+      post.custom_fields[DiscourseTranslator::DETECTED_LANG_CUSTOM_FIELD] ||= client.translate_text({
+        text: post.cooked.truncate(MAXLENGTH, omission: nil),
+        source_language_code: 'auto',
+        target_language_code: SUPPORTED_LANG[I18n.locale],
+      })&.source_language_code
+    end
+
+    def self.translate(post)
+      from_custom_fields(post) do
+        result = client.translate_text({
+          text: post.cooked.truncate(MAXLENGTH, omission: nil),
+          source_language_code: "auto",
+          target_language_code: SUPPORTED_LANG[I18n.locale],
+        })
+
+        detected_lang = post.custom_fields[DiscourseTranslator::DETECTED_LANG_CUSTOM_FIELD] ||= result.source_language_code
+
+        [detected_lang, result.translated_text]
+      end
+    rescue Aws::Translate::Errors::UnsupportedLanguagePairException
+      raise I18n.t('translator.failed')
+    end
+
+    def self.client
+      @credentials ||= Aws::Credentials.new(SiteSetting.translator_aws_key_id, SiteSetting.translator_aws_secret_access)
```

Comment:
We need a small change here, if the key ID and secret are not provided, we shouldn't pass a `credentials` parameter to `Translate::Client` that way, the API can try loading the credentials automatically. See https://docs.aws.amazon.com/sdk-for-ruby/v2/api/Aws/Translate/Client.html